Support global and per endpoint / backend static attributes in all layers

ssepml opened the pull request:
## Description

This PR adds support for:

- Global static attributes for `global` and `proxy` layers
- Set per endpoint / backend static attributes. These will be merged with the globally provided ones if any

## Use Case

The main idea is to provide allow for better grouping when creating dashboards with the metrics.

## Approach

Two new fields were added to the `global` and `proxy` layers

- `metrics_static_attributes`
- `traces_static_attributes`

The reason for this option was that the configuration on these two layers was already flat (as opposed to the `backend` layer which has a `metrics` and `traces` section) , hence this keeps the convention already set. Also, changing to something like the `backend` layer could lead to some breaking changes.

__NOTE__: I don't have strong opinions on this format and I'm willing to adapt the PR to your preferred way to pass the configuration

Also, configuration provided in an endpoint of backend section will now be parsed to get static labels which will be merged with any other label provided at the top level of the configuration. All other configuration parameters provided will be ignored.

ssepml committed:
Basic implemetation of static attributes in all layers

ssepml committed:
Split metrics and traces attributes

ssepml committed:
Extend static attributes override to proxy and backend layers

ssepml committed:
Added new static labels examples to backend example server config

ssepml committed:
Added some testing

dhontecillas commented on state/config.go:
```diff
@@ -1,22 +1,21 @@
 	EndpointOTEL(cfg *luraconfig.EndpointConfig) OTEL
-	EndpointPipeOpts(cfg *luraconfig.EndpointConfig) *config.PipeOpts
-	EndpointBackendOpts(cfg *luraconfig.Backend) *config.BackendOpts
+	EndpointPipeOpts(cfg *luraconfig.EndpointConfig) config.PipeOpts
+	EndpointBackendOpts(cfg *luraconfig.Backend) config.BackendOpts
```
the change of the signature from `*config.PipeOpts` to `config.PipeOpts` is a breaking backwards compat change, and should be avoided.

dhontecillas commented on state/config.go:
```diff
@@ -32,28 +31,65 @@ func (*StateConfig) OTEL() OTEL {
+}
+
+func mergedBackendOpts(s *StateConfig, cfg *luraconfig.Backend) config.BackendOpts {
+	BackendOpts := *s.cfgData.Layers.Backend
```
I would say variables should not start with a capital character.

dhontecillas commented on state/config.go:
```diff
@@ -32,28 +31,65 @@ func (*StateConfig) OTEL() OTEL {
+	return mergedBackendOpts(s, cfg)
+}
+
+func mergedBackendOpts(s *StateConfig, cfg *luraconfig.Backend) config.BackendOpts {
```
I would say this function could be a "member" of s , and be:
```func (s *StateConfig) mergedBackendOpts(cfg *luraconfig.Backend) *config.BackendOpts {```

dhontecillas commented on config/lura.go:
```diff
@@ -46,3 +40,24 @@ func FromLura(srvCfg luraconfig.ServiceConfig) (*ConfigData, error) {
+
+// LuraExtraCfg extracts the extra config field for the namespace if
+// provided
+func LuraExtraCfg(extraCfg luraconfig.ExtraConfig) (*ConfigData, error) {
```
this makes the code more readable ! :+1:

dhontecillas commented on example/Makefile:
```diff
@@ -1,8 +1,8 @@
+srv:
 	go build -o srv ./server
-.PHONY: build
+.PHONY: srv
```
i think, we can remove the `.PHONY` line if we actually generate a file with the target name.

dhontecillas reviewed (changes requested):
I took your PR as base and put it into this other branch: https://github.com/krakend/krakend-otel/pull/27 .

I would like to make a PR to your branch, so you could review it, and then merge this PR. But since I merged your PR in the main repo (I guess I made a little mess), I don't know how to do it. So, would you mind reviewing that one ?

Luckily it maintains you as an author of the commits (I did not want to take credit for your work, just tried to test it in existing code bases)

Another way would be ... to .. merge that branch into yours ? 

Sorry for the mess. 

And thank you very much for your contribution.

ssepml commented:
@dhontecillas thanks for the comments. I reviewed the other PR you created, all looks good to me. 

> Luckily it maintains you as an author of the commits (I did not want to take credit for your work, just tried to test it in existing code bases)

Don't worry about that :), I don't really mind which PR gets merged and given that your PR addresses all the comments already, feel free to continue with that one.

dhontecillas commented:
@ssepml then I will merge the other one directly.

Once again, thank you very much for your contribution.

This PR is merged into: https://github.com/krakend/krakend-otel/pull/27
